### PR TITLE
Overflow- and Dust-Aware Box Selector

### PR DESCRIPTION
**Pull request by kushti**
This PR is introducing a box selector which is parameterized by maximum number of inputs a transaction can have, and optimal number of inputs.

Say, the selector is given boxes denoted by their values (1,2,3,4,...10). Then the selector is working as follows:

    1) the selector first picking up boxes in given order (1,2,3,4,...)
    2) if number of inputs exceeds the limit, the selector is sorting remaining boxes(actually, only 3*maximum inputs of them) by value in descending order and replaces small-value boxes in the inputs by big-value from the tail (1,2,3,4 => 10)
    3) if the number of inputs still exceeds the limit, the selector is trying to throw away the dust if possible. E.g. if inputs are (100, 200, 1, 2, 1000), target value is 1300 and maximum number of inputs is 3, the selector kicks out (1, 2)
     4) if number of inputs after the previous steps is below optimal, the selector is trying to append the dust, by sorting remaining boxes in ascending order and appending them till optimal number of inputs.

### CODE REVIEW

File: src/main/scala/org/ergoplatform/wallet/boxes/ReplaceCompactCollectBoxSelector.scala [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    * @return None if select() is failing to pick appropriate boxes, otherwise Some(res), where res contains boxes
+    *         to spend as well as monetary values and assets for boxes containing change
+    *         (wrapped in a special BoxSelectionResult class).
+    */
+  override def select(inputBoxes: Iterator[TrackedBox],
+                      filterFn: TrackedBox => Boolean,
+                      targetBalance: Long,
+                      targetAssets: Map[ModifierId, Long]): Option[BoxSelector.BoxSelectionResult] = {
+    DefaultBoxSelector.select(inputBoxes, filterFn, targetBalance, targetAssets).flatMap { initialSelection =>
+      val tail = inputBoxes.take(maxInputs * ScanDepthFactor).filter(filterFn).toSeq
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/wallet/boxes/ReplaceCompactCollectBoxSelector.scala]: We may miss some useful boxes here

File: src/main/scala/org/ergoplatform/wallet/boxes/ReplaceCompactCollectBoxSelector.scala [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        if (afterCompaction.boxes.length > maxInputs) {
+          None
+        } else if (afterCompaction.boxes.length < optimalInputs) {
+          collectDust(afterCompaction, tail, targetBalance, targetAssets)
+        } else Some(afterCompaction)
+      }
+    }
+  }
+
+  def collectDust(bsr: BoxSelectionResult,
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/wallet/boxes/ReplaceCompactCollectBoxSelector.scala]: please add comments to public methods

Comment (kushti) [reply src/main/scala/org/ergoplatform/wallet/boxes/ReplaceCompactCollectBoxSelector.scala]: made the inner methods protected

File: src/main/scala/org/ergoplatform/wallet/boxes/ReplaceCompactCollectBoxSelector.scala [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+  * 3) if the number of inputs still exceeds the limit, the selector is trying to throw away the dust if possible. E.g. if inputs are (100, 200, 1, 2, 1000), target value is 1300 and maximum number of inputs is 3, the selector kicks out (1, 2)
+  * 4) if number of inputs after the previous steps is below optimal, the selector is trying to append the dust, by sorting remaining boxes in ascending order and appending them till optimal number of inputs.
+  *
+  * @param maxInputs     - maximum number of inputs a transaction can have
+  * @param optimalInputs - optimal number of inputs, when transaction is still not expensive. The box selector is
+  *                      trying to add dust if a transaction has less inputs than this.
+  */
+class ReplaceCompactCollectBoxSelector(maxInputs: Int, optimalInputs: Int) extends BoxSelector {
+
+  val ScanDepthFactor = 4
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/wallet/boxes/ReplaceCompactCollectBoxSelector.scala]: Potentially, if there are many boxes (e.g. thousands), the tail processing would become slow.

**Pull request merged**